### DISCUSSION TITLE
Update on my program

### DISCUSSION THREAD

**Discussion by texdraft** [shreevatsa/webWEB#17] [Show and tell]
Hi all. I wanted to give an update on the progress on my program. Currently it's able to handle formatting vanilla Pascal pretty well, but not `WEB`. I've put the [output when run on P4](https://texdraft.github.io/p4/program) on my GitHub pages. Some caveats:

- I've removed all the comments. They didn't look good and messed up the formatting in places. As stated in the [thirteenth issue of the Pascal Newsletter, page 48](http://pascal.hansotten.com/uploads/pug/13_Pascal_Newsletter_Dec78.pdf#page=50), “The bane of most Pascal prettyprinters is the treatment of comments”. They can confuse `WEAVE` too.
- The standalone “meaning pages” (which you get to if you click on a defining/declaring occurrence of an identifier) aren't very useful. All they do is let you see their declaration and/or definition. (For standard identifiers, they are empty.) Ideally they'd also contain all the places where the identifier is referred to, but adding this is proving to be tricky.

If you read the HTML, you might notice that `<div>` elements are used for indentation, rather than `<table>` elements. HTML tables are strange things and are hard to control. Switching to `<div>`s halved the depth of the DOM (`<table><tbody><tr><td>` vs `<div><div>`) and simplified styling. Furthermore, [browsers place limits on nested tables](https://stackoverflow.com/questions/39969762). You might also notice that there is no indentation in the HTML. The reason is that indentation significantly inflates the size of the files, which are already large enough. Other than the lack of `<table>`s and indentation, the generated output is a lot like the Common Words proof of concept I posted a while ago. (As I write this I notice a bug; the `program` line should of course be in the same `<div class="code">` as the rest of the markup. Oh well.)

The formatting is configurable. In the uploaded output all of the default settings are used. To see what the page looks like with your system's default style and fonts, add the `disabled` attribute to the line

```
<link href="extra.css" rel="stylesheet">
```

in the HTML files.

I'm interested in hearing any feedback about reading and navigating the document. (For a real `WEB` program there will be a table of contents etc.) Also, what should the index look like? Should entries for identifiers be organized by meaning at all (and take you to the page for the meaning, rather than to a section), or should they be as in woven output?

I'd prefer to update my repository only once the program can format `WEB` files without dying, but it's no problem if anyone wants to see the code.

**Comment by shreevatsa** [comment]
_(Sorry I couldn't reply earlier; things are bad at work…)_

This is cool, thanks for the update!

A couple of comments:

- On navigation: Given the deeply nested nature of this code, it would be nice to be able to go from the start of a block to its end. Otherwise it's hard to keep track of indentation visually, to go to the next procedure if no longer interested in the current one, etc. I imagine the simplest implementation would be to have clicking on each `begin` to jump to its corresponding `end`. (More fancy would be expand-collapse with +/- buttons or whatever.)

- On updating the repository: I don't know about you, but for me my time-spent on various things seems to go up and down randomly, so for any given partially-completed things there's a good chance I may never return to it again. :-( In such a situation, it would be better to have the current state dumped somewhere, so that at least others can make use of it. (There are many practically abandoned projects of mine where I didn't follow this, and now I can't even find the corresponding code or notes or whatever.) So if it were me I'd prefer to err on the side of updating frequently—if the concern is about having a clean commit history, note that you can always have the in-progress work in a separate branch, or just rewrite history with a force-push later. :-)

Thanks for this, it's really cool to see the progress here,
